Contributions/Linux_Kernel_Safety_First_Principles.md

```diff
@@ -0,0 +1,141 @@
+
+[First Principles - Availability](#First-Principles---Availability)
+
+[Considerations - Toward the safety goals](#Considerations---Toward-the-safety-goals)
+
+[License: CC BY-SA 4.0](#License-CC-BY-SA-40)
+
+## **Terms and Abbreviations**
+
+   See the matching section in *Interference Scenarios for an ARM64 Linux System*
+
+## **References**
+
+1. ***Interference Scenarios for an ARM64 Linux System***
+2. ***Safety Requirements for a Generic Linux System***
+3. ***CheckList for Safety Claims on a Generic Linux System***
+4. ***Using Linux in a Safe System***
+5. ***Linux Memory Management Essentials***
+6. [CC BY-SA 4.0 Deed | Attribution-ShareAlike 4.0 International | Creative Commons](https://creativecommons.org/licenses/by-sa/4.0/) - <https://creativecommons.org/licenses/by-sa/4.0/> License
+
+
+## **Disclaimer**
+This is not intended to be a complete list of facts, however it attempts
+to summarise the most relevant ones from the documents listed in the
+References section.
+
+## **Purpose of the document**
+Summarise the concepts from the documents in the Reference sections.
+Justifications for each statement can be found throughout those documents.
+
+## **Structure of the document**
+The next section illustrates the philosohpy behind the approach used here to
+deal with safety requirements vs a system using Linux.
+The further two sections about "First Principles" address two aspects of safety:
+-  integrity: reliably detecting interference to components with safety requirements
+-  availability: protracted integrity of the safey-relevant functionality
+The last section lists considerations that can help with raching the safety goals.
+
+
+## **Philosophy of the Document**
+*  A system subject to safety goals might employ the Linux Kernel.
+*  Safety goals are specific to the overall system and the related use cases.
+*  Similarly, one must expect that the solution will be tailored to both system and goals.
+*  However, it is possible to describe the Linux Kernel from a safety point of view.
+*  Not all the issues listed will be necessarily relevant to each system/application.
+*  A Safety Analysis should, however, clearly state which issues are relevant, which are not, and why.
+
+
+## **First Principles - Integrity**
+This section provides touchstone concepts about interference of QM components with
+safety-qualified ones.
+
+1. The vanilla Linux Kernel, is not sufficient to support, alone, safey goals.
+2. No internal kernel barriers/protections against interference to variable data, including safety-relevant data.
+3. Any component within the kernel context can generate (cascaded) interference.
+4. Internal interference can cascade also through components that have been qualified for safety at unit-level.
+5. The Linux Kernel is able to interfere with both itself and any part of user-space processes (linear map).
+6. No generic solution for very complex systematic corruption of any writable memory.
```

Comment:
Here what do you mean by "generic solution"? Do you mean "currently there is no design solution to protect writable safety relevant memory"?

Reply:
"Currently" is implied, but, again, as long as something is proven to work, it's an absolute statement.
It can be rephrased, if it helps with readability, without altering the meaning.
My point is that the "Very complex systematic corruption" cannot be detected through solutions that are applicable to - let's say - the memory map of a safety-relevant process, that might comprise tens, hundreds of MB of memory. You can certainly try to use ad-hoc verification, e.g. performing functional testing on that process, but it needs to be tailored to each individual process. No generic solution. Only a generic methodology, to be taylored.

Reply:
I think we can make a stronger (but narrower) statement here:
```suggestion
6. Complex systematic corruption of writable memory cannot be prevented by a generic, kernel-level solution
```

Reply:
@reiterative: "neither prevented nor detected" is the right formualtion, I think

Reply:
@igor-stoppa  I think your statement is not correct. For example CONFIG_STACKPROTECTOR_STRONG or KASAN are generic Kernel solution that provide detections of writable memory. So I would stick with Pauls' rewording

Reply:
@paolonig They are partial solutions. Placing canaries on the stack doesn't catch all the possible corruptions.
KASAN alters the system configuration, and anyways is not deployed.
Feel free to provide evidence countering my arguments, they are intentionally made in a very clan cut way, to allow easy confutation by evidence, if any can be found.

Reply:
Yes they are partial generic kernel level solutions, I agree. However an integrator or a SW vendor may use such partial solution in combination with other measures to support a non-interference claim.
So IMO the fact is the monolithic nature of the Kernel that does not allow, today, interference prevention (like the one you can have by SMMU for a user space process) and we can all agree on this.
What I cannot agree is about principles that also imply or deny specific ways to support a certain claim; since the specific ways depend on the specification of the integration context and of other projects aspects (depending on the engineering workforce I may decide to use some verification strategies rather than others....)